sklearn/preprocessing/_target_encoder.py

```diff
@@ -0,0 +1,304 @@
+        "cv": ["cv_object"],
+    }
+
+    def __init__(self, categories="auto", target_type="auto", smooth="auto", cv=5):
+        self.categories = categories
+        self.smooth = smooth
+        self.target_type = target_type
+        self.cv = cv
+
+    def fit(self, X, y):
+        """Fit the :class:`TargetEncoder` to X and y.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data to determine the categories of each feature.
+
+        y : array-like of shape (n_samples,)
+            The target data used to encode the categories.
+
+        Returns
+        -------
+        self : object
+            Fitted encoder.
+        """
+        self._validate_params()
+        self._fit_encodings_all(X, y)
+        return self
+
+    def fit_transform(self, X, y):
+        """Fit :class:`TargetEncoder` and transform X with the target encoding.
+
+        .. note::
+            :class:`TargetEncoder` uses a cross validation scheme in
+            :meth:`fit_transform` to prevent leaking the target during training. In
+            :meth:`fit_transform`, the data is split according to the `cv` parameter.
+            Categorical encodings are learned from split and used to encode the other
+            split. Afterwards, a final categorical encoding is learned from all the
+            data, which is then used to encode data during :meth:`transform`. This
+            means that `fit(X, y).transform(X)` does not equal `fit_transform(X, y)`.
+
+        Parameters
+        ----------
+        X : array-like of shape (n_samples, n_features)
+            The data to determine the categories of each feature.
+
+        y : array-like of shape (n_samples,)
+            The target data used to encode the categories.
+
+        Returns
+        -------
+        X_trans : ndarray of shape (n_samples, n_features)
+            Transformed input.
+        """
+        from ..model_selection._split import check_cv  # avoid circular import
+
+        self._validate_params()
+        X_ordinal, X_known_mask, y, n_categories = self._fit_encodings_all(X, y)
+        cv = check_cv(self.cv)
+        X_out = np.empty_like(X_ordinal, dtype=np.float64)
+        X_unknown_mask = ~X_known_mask
+
+        for train_idx, test_idx in cv.split(X, y):
+            X_train, y_train = X_ordinal[train_idx, :], y[train_idx]
+            y_mean = np.mean(y_train)
+
+            if self.smooth == "auto":
+                y_variance = np.var(y_train)
+                encodings = _fit_encoding_fast_auto_smooth(
+                    X_train, y_train, n_categories, y_mean, y_variance
+                )
+            else:
+                encodings = _fit_encoding_fast(
+                    X_train, y_train, n_categories, self.smooth, y_mean
+                )
+            self._transform_X_ordinal(
+                X_out, X_ordinal, X_unknown_mask, test_idx, encodings, y_mean
+            )
```

Comment:
In case of overlapping `test_idx`, e.g. when using `ShuffleSplit` with a larger number of splits, the last iterations will overwrite the previously encoded values.

Is this a problem, should we try to detect this case and warn the user? Or just document in the `cv` parameter.

What about cases where not all samples appear at lest once as `test_idx`? E.g. `cv=ShuffleSplit(n_splits=1, test_size=0.5)`? Reading this code I would have thought that `X_out` would hold uninitialized values (from the `np.empty_like` allocation above). However it does not seem to be the case when I tried but I am not sure I understand why...

Reply:
Actually, it can show it is a problem when disabling smoothing:

```python
>>> from sklearn.model_selection import ShuffleSplit
>>> from sklearn.preprocessing import TargetEncoder
>>> import numpy as np
>>> X = np.random.randint(0, 3, 30).reshape(-1, 1)
>>> TargetEncoder(cv=ShuffleSplit(n_splits=1, test_size=0.5), smooth=0).fit_transform(X, X.ravel() + np.random.randn(X.shape[0]))
array([[ 0.00000000e+000],
       [ 9.88131292e-324],
       [ 4.22543968e-001],
       [ 9.88131292e-324],
       [ 0.00000000e+000],
       [ 4.22543968e-001],
       [ 9.88131292e-324],
       [ 4.94065646e-324],
       [ 4.94065646e-324],
       [ 2.26479250e+000],
       [ 2.26479250e+000],
       [ 4.22543968e-001],
       [ 0.00000000e+000],
       [ 4.22543968e-001],
       [ 4.22543968e-001],
       [-5.53085205e-002],
       [-5.53085205e-002],
       [ 2.26479250e+000],
       [ 4.94065646e-324],
       [-5.53085205e-002],
       [ 4.94065646e-324],
       [ 4.22543968e-001],
       [ 4.22543968e-001],
       [ 4.94065646e-324],
       [ 9.88131292e-324],
       [ 2.26479250e+000],
       [ 2.26479250e+000],
       [ 9.88131292e-324],
       [ 9.88131292e-324],
       [ 0.00000000e+000]])
```

Reply:
Those `e-324` values look fishy.

Reply:
In this case, I think it's best to only allow integers for the `cv` parameter for the initial version. As you said, there are quite a few ways to get the wrong result with a custom CV.

Reply:
Fair enough, an alternative would be to check for non-overlapping validation folds but this can be expensive. Let's start with integer specified k-fold and we will see later if we need to expand the allowed parametrization for the `cv` parameter or not. But I agree this is probably a YAGNI.